File: src/gestorAplicación/interno/Corriente.java
package gestorAplicación.interno;

import gestorAplicación.externo.Banco;
import gestorAplicación.externo.Cuotas;
import gestorAplicación.externo.Divisas;
import java.util.ArrayList;
import java.util.Date;
import java.time.Instant;

public class Corriente extends Cuenta{
	//Atributos
	private static final long serialVersionUID = 8L;
	public static final String nombreD = "Corriente";
	private double cupo;
	private double disponible;
	private Cuotas plazo_Pago;
	//Tasa efectiva anual
	private double intereses;
	//Atributo que decide pago de interes en primer mes
	private boolean primerMensualidad;
	private static ArrayList<Corriente> cuentasCorrienteTotales = new ArrayList<Corriente>();;
	
	
	//Constructores
	//Hacer chequeo, cupo viene por defecto según suscripción y banco asociado.
	public Corriente(Banco banco, int clave, Divisas divisa, String nombre) {
		super(banco, clave, divisa, nombre);
		Corriente.getCuentasCorrienteTotales().add(this);
	}	
	
	public Corriente(Banco banco, int clave, String nombre) {
		super(banco, clave, nombre);
		Corriente.getCuentasCorrienteTotales().add(this);
	}
	
	public Corriente() {
		super();
		Corriente.getCuentasCorrienteTotales().add(this);
	}
	
	//Métodos
	public Corriente crearCuenta(Banco banco, int clave, Divisas divisa, String nombre) {
		return (new Corriente(banco, clave, divisa, nombre));
	}
	
	public Corriente crearCuenta(Banco banco, int clave, String nombre) {
		return (new Corriente(banco, clave, nombre));
	}
	
	public double[] retornoCuotaMensual(double DeudaActual) {
		double[] cuotaMensual = new double[3];
		double interes_nominal_mensual = Corriente.calculoInteresNominalMensual(this.getIntereses());
		double interes = DeudaActual * (interes_nominal_mensual / 100);
		cuotaMensual[0] = interes;
		double abono_capital = this.getDisponible() / this.getPlazo_Pago().getCantidad_Cuotas();
		cuotaMensual[1] = abono_capital;
		double cuotaMensualFinal = interes + abono_capital;
		cuotaMensual[2] = cuotaMensualFinal;
		return cuotaMensual;
	}
	
	public double[] retornoCuotaMensual(double DeudaActual, int mes) {
		double[] cuotaMensual = new double[3];
		double interes_nominal_mensual = Corriente.calculoInteresNominalMensual(this.getIntereses());
		if(mes == 1) {
			cuotaMensual[0] = 0;
			double abono_capital = this.getDisponible() / this.getPlazo_Pago().getCantidad_Cuotas();
			cuotaMensual[1] = abono_capital;
			double cuotaMensualFinal = abono_capital;
			cuotaMensual[2] = cuotaMensualFinal;
		}
		else if(mes == 2){
			double abono_capital = this.getDisponible() / this.getPlazo_Pago().getCantidad_Cuotas();
			double interes_mes1 = (interes_nominal_mensual / 100) * (abono_capital + DeudaActual);
			double interes_mes2 = DeudaActual * (interes_nominal_mensual / 100);
			double interes = interes_mes1 + interes_mes2;
			cuotaMensual[0] = interes;
			cuotaMensual[1] = abono_capital;
			double cuotaMensualFinal = interes + abono_capital;
			cuotaMensual[2] = cuotaMensualFinal;
		}
		
		return cuotaMensual;
	}
	
	public static String imprimirCuotaMensual(double[] cuotaMensual) {
		return "Cuota: " + cuotaMensual[2] +
				"\nIntereses: " + cuotaMensual[1] +
				"\nAbono a capital: " + cuotaMensual[0];
	}
	
	public static double calculoInteresNominalMensual(double interesEfectivoAnual) {
		double interes = Math.pow((1 + interesEfectivoAnual), (30 / 360)) - 1;
		return interes;
	}
	
	// Funcionalidad asesor inversiones

	public void vaciarCuenta(Ahorros gota) {
		Movimientos movimiento = new Movimientos(this, gota, this.getDisponible(), Categoria.OTROS,
				Date.from(Instant.now()));
		this.getTitular().asociarMovimiento(movimiento);
	}
	
	//Funcionalidad Compra de Cartera
	public static Corriente vistaPreviaMovimiento(Corriente cuenta, Cuotas plazo, double Deuda_previa, double interes) {
		Corriente cuenta_aux = cuenta;
		cuenta_aux.setDisponible(cuenta.getDisponible() - Deuda_previa);
		cuenta_aux.setIntereses(interes);
		cuenta_aux.setPlazo_Pago(plazo);
		return cuenta_aux;
	}
	
	//Método creado como calculadora de cuotas mensuales para pago de un préstamo
		//El atributo interes hace referencia a la tasa efectiva anual
	public static double[][] calculadoraCuotas(Cuotas cuotas, double deuda, double intereses) {
		int cuotasTotales = cuotas.getCantidad_Cuotas();
		double[][] cuota = new double[cuotasTotales][3];
		double interesMensual = Corriente.calculoInteresNominalMensual(intereses);
		double deudaActual = deuda;
		
		double abono_capital = deuda / cuotasTotales;
		
		for (int i = 0; i <= cuotasTotales; i++) {
			double[] cuotaMes = new double[3];
			double interes = deudaActual * (interesMensual / 100);
			cuotaMes[0] = interes;
			double cuota_pagar = interes + abono_capital;
			cuotaMes[1] = cuota_pagar;
			double deudaTotal = deudaActual - (cuota_pagar - interes);
			cuotaMes[2] = deudaTotal;
			cuota[i] = cuotaMes;
			
			deudaActual = deudaTotal;
		}
		
		return cuota;
	}
	
	public static double[][] calculadoraCuotas(Cuotas cuotas, double deuda, double intereses, boolean auxiliar){
		int cuotasTotales = cuotas.getCantidad_Cuotas();
		double[][] cuota = new double[cuotasTotales][3];
		double interesMensual = Corriente.calculoInteresNominalMensual(intereses);
		double deudaActual = deuda;
		
		double abono_capital = deuda / cuotasTotales;
		
		double interesMes1 = deudaActual * (interesMensual / 100);
		double[] cuotaMes1 = new double[3];
		cuotaMes1[0] = 0;
		cuotaMes1[1] = abono_capital;
		cuotaMes1[2] = deudaActual - abono_capital;
		cuota[0] = cuotaMes1;
		
		deudaActual = deudaActual - abono_capital;
		
		for (int i = 1; i <= cuotasTotales; i++) {
			double[] cuotaMes = new double[3];
			double interes = deudaActual * (interesMensual / 100);
			cuotaMes[0] = interes;
			double cuota_pagar = interes + abono_capital + interesMes1;
			cuotaMes[1] = cuota_pagar;
			double deudaTotal = deudaActual - (cuota_pagar - interes);
			cuotaMes[2] = deudaTotal;
			cuota[i] = cuotaMes;
			
			interesMes1 = 0;
			deudaActual = deudaTotal;
		}
		
		return cuota;
	}
	
	public static double[] informacionAdicionalCalculadora (double[][] cuota, double deuda) {
		double[] infoAdicional = new double[3];
		double totalPagado = 0;
		
		for (int i = 0; i <= cuota.length; i++) {
			totalPagado += cuota[i][1];
		}
		
		double interesesPagados = totalPagado - deuda;
		
		infoAdicional[0] = totalPagado;
		infoAdicional[1] = interesesPagados;
		infoAdicional[2] = deuda;
		
		return infoAdicional;
	}
	
	public int compareTo(Corriente cuenta) {
		if(this.getDisponible() > cuenta.getDisponible()) {
			return 1;
		}
		else if(this.getDisponible() < cuenta.getDisponible()) {
			return -1;
		}
		else {
			return 0;
		}
	}
	
	public Double getCupo() {
		return cupo;
	}
	
	public void setCupo(Double cupo) {
		this.cupo = cupo;
	}
	
	public Cuotas getPlazo_Pago() {
		return plazo_Pago;
	}
	
	public void setPlazo_Pago(Cuotas plazo_Pago) {
		this.plazo_Pago = plazo_Pago;
	}
	
	public Double getDisponible() {
		return disponible;
	}
	
	public void setDisponible(Double disponible) {
		this.disponible = disponible;
	}

	public double getIntereses() {
		return intereses;
	}

	public void setIntereses(double intereses) {
		this.intereses = intereses;
	}

	public boolean getPrimerMensualidad() {
		return primerMensualidad;
	}
	
	public void setPrimerMensualidad(boolean primerMensualidad) {
		this.primerMensualidad = primerMensualidad;
	}
	
	public String toString() { 
		return "Cuenta: " + this.nombre +
				"\nCuenta Corriente # " + this.id +
				"\nBanco: " + this.banco +
				"\nDivisa: " + this.divisa +
				"\nCupo disponible: " + this.cupo + " " + this.divisa +
				"\nCuotas: " + this.plazo_Pago;
	}
	
	public static ArrayList<Corriente> getCuentasCorrienteTotales() {
		return cuentasCorrienteTotales;
	}

	public static void setCuentasCorrienteTotales(ArrayList<Corriente> cuentasCorrienteTotales) {
		Corriente.cuentasCorrienteTotales = cuentasCorrienteTotales;
	}
}

File: Python/src/gestorAplicación/interno/corriente.py
from .cuenta import Cuenta
from gestorAplicación.externo.banco import Banco
from datetime import date
import math

class Corriente(Cuenta):
    #Atributos de clase
    _cuentasCorrienteTotales = []

    #Constructor
    def __init__(self, banco, clave, divisa, nombre):
        #Atributos de instancia
        Corriente._cuentasCorrienteTotales.append(self)
        super().__init__(banco, clave, divisa, nombre)

#REVISAR SOBRECARGA

    def crearCuenta(self, banco, clave, divisa, nombre):
        return Corriente(banco, clave, divisa, nombre)
    
    def __str__(self):
        return "Cuenta: " + self._nombre + "\nCuenta Corriente # " + self._id + "\nTitular: " + self.getTitular().getNombre() + "\nBanco: " + self._banco.getNombre() + "\nDivisa: " + self._divisa + "\nCupo: " + self._cupo + " " + self._divisa + "\nCupo disponible: " + self._disponible + " " + self._divisa + "\Cuotas: " + self._plazo_Pago + "\nIntereses: " + self._intereses
    
    # Método para la funcionalidad asesoramiento de inversiones
    def vaciarCuenta(self, gota):
        from .movimientos import Movimientos
        from .categoria import Categoria
        movimiento = Movimientos(self, gota, self.getDisponible(), Categoria.OTROS, date.now())
        self.getTitular().getMovimientosAsociados().append(movimiento)
        Movimientos.getMovimientosTotales().remove(movimiento)

    def retornoCuotaMensual(self, deudaActual, mes = 0):
        cuotaMensual = []
        interes_nominal_mensual = Corriente.calculoInteresNominalMensual(self.getIntereses())
        if mes == 0:
            interes = deudaActual * (interes_nominal_mensual / 100)
            cuotaMensual[0] = interes
            abono_capital = (self.getCupo() - self.getDisponible()) / self.getPlazo_Pago().getCantidad_Cuotas()
            cuotaMensual[1] = abono_capital
            cuotaMensualFinal = interes + abono_capital
            cuotaMensual[2] = cuotaMensualFinal
        elif mes == 1:
            cuotaMensual[0] = 0
            abono_capital = (self.getCupo() - self.getDisponible()) / self.getPlazo_Pago().getCantidad_Cuotas()
            cuotaMensualFinal = abono_capital
            cuotaMensual[2] = cuotaMensualFinal
        elif mes == 2:
            abono_capital = (self.getCupo() - self.getDisponible()) / self.getPlazo_Pago().getCantidad_Cuotas()
            interes_mes1 = (interes_nominal_mensual / 100) * (abono_capital + deudaActual)
            interes_mes2 = deudaActual * (interes_nominal_mensual / 100)
            interes = interes_mes1 + interes_mes2
            cuotaMensual[0] = interes
            cuotaMensual[1] = abono_capital
            cuotaMensualFinal = interes + abono_capital
            cuotaMensual[2] = cuotaMensualFinal
        else:
            interes = deudaActual * (interes_nominal_mensual / 100)
            cuotaMensual[0] = interes
            abono_capital = (self.getCupo() - self.getDisponible()) / self.getPlazo_Pago().getCantidad_Cuotas()
            cuotaMensual[1] = abono_capital
            cuotaMensualFinal = interes + abono_capital
            cuotaMensual[2] = cuotaMensualFinal
        return cuotaMensual

    @staticmethod
    def imprimitCuotaMensual(cls, cuotaMensual):
        return "Cuota: " + Corriente.redondeoDeciomal(cuotaMensual[2], 2) + "\nIntereses: " + Corriente.redondeoDecimal(cuotaMensual[0], 2) + "\nAbono a capital: " + Corriente.redondeoDecimal(cuotaMensual[1], 2)

    @staticmethod
    def calculoInteresNominalMensual(cls, interesEfectivoAnual):
        interes = math.pow((1 + (interesEfectivoAnual / 100)), (30.0 / 360.0)) - 1
        interes_porcentaje = interes * 100
        interes_porcentaje_redondeado = Corriente.redondeoDecimal(interes_porcentaje, 2)
        return interes_porcentaje_redondeado

    @staticmethod
    def vistaPreviaMovimiento(cls, cuenta, plazo, deuda_Previa, interes):
        cuenta_aux = cuenta.clone()
        cuenta_aux.setDisponible(cuenta.getDisponible() - deuda_Previa)
        cuenta_aux.setIntereses(interes)
        cuenta_aux.setPlazo_Pago(plazo)
        return cuenta_aux
    #Revisar necesidad del clonado

    @staticmethod
    def calculadoraCuotas(cls, cuotas, deuda, intereses, auxiliar):
        cuotasTotales = cuotas.getCantidad_Cuotas()
        cuota = []
        interesMensual = Corriente.calculoInteresNominalMensual(intereses)
        deudaActual = deuda

        abono_capital = deuda / cuotasTotales

        if not auxiliar:
            for i in range(0, i < cuotasTotales, 1):
                cuotaMes = []
                interes = deudaActual * (interesMensual / 100)
                cuotaMes[0] = interes
                cuota_pagar = interes + abono_capital
                cuotaMes[1] = cuota_pagar
                deudaTotal = deudaActual - (cuota_pagar - interes)
                cuotaMes[2] = deudaTotal
                cuota[i] = cuotaMes

                deudaActual = deudaTotal
        else:
            interesMes1 = deudaActual * (interesMensual / 100)
            cuotaMes1 = []
            cuotaMes1[0] = 0
            cuotaMes1[1] = abono_capital
            cuotaMes1[2] = deudaActual - abono_capital
            cuota[0] = cuotaMes1

            deudaActual = deudaActual - abono_capital

            for i in range(1, 1 < cuotasTotales, 1):
                cuotaMes = []
                interes = deudaActual * (interesMensual / 100)
                cuotaMes[0] = interes + interesMes1
                cuota_pagar = interes + abono_capital + interesMes1
                cuotaMes[1] = cuota_pagar
                deudaTotal = deudaActual - (cuota_pagar - (interes + interesMes1))
                cuotaMes[2] = deudaTotal
                cuota[i] = cuotaMes

                interesMes1 = 0
                deudaActual = deudaTotal
            
        return cuota
            
        

    @staticmethod
    def informacionAdicionalCalculadora(cls, cuota, deuda):
        infoAdicional = []
        totalPagado = 0

        for i in range(0, len(cuota), 1):
            totalPagado += cuota[i][1]
        
        interesesPagados = totalPagado - deuda

        totalPagado = Corriente.redondeoDecimal(totalPagado, 2)
        interesesPagados = Corriente.redondeoDecimal(interesesPagados, 2)
        deuda = Corriente.redondeoDecimal(deuda, 2)

        infoAdicional[0] = totalPagado
        infoAdicional[1] = interesesPagados
        infoAdicional[2] = deuda

        return infoAdicional

    @staticmethod
    def inicializarCupo(cls, cuenta):
        banco = cuenta.getBanco()
        suscripcion = cuenta.getTitular().getSuscripcion()
        cupo = Banco.decisionCupo(suscripcion, banco)
        cuenta.setCupo(cupo)
        cuenta.setDisponible(cupo)

    def capacidadDeuda(self, cantidad):
        if self.getDisponible().compareTo(cantidad) >= 0:
            return True
        else:
            return False

    @staticmethod
    def limpiarPropiedades(cls, arreglo):
        arreglo.remove("cuentasTotales")
        #Verificar que otras variables se crean
    #Verificar su uso

    #Realizar el método del compare to
    def compareTo(self, cuenta):
        if self.getDisponible() > cuenta.getDisponible():
            return 1
        elif self.getDisponible() < cuenta.getDisponible():
            return -1
        else:
            return 0

    #Revisar la parte del clone
    

    #Atributos Get & Set
    def getCupo(self):
        return self._cupo
    def setCupo(self, cupo):
        self._cupo = cupo

    def getPlazo_Pago(self):
        return self._plazo_Pago
    def setPlazo_Pago(self, plazo_Pago):
        self._plazo_Pago = plazo_Pago
    
    def getDisponible(self):
        return self._disponible
    def setDisponible(self, disponible):
        self._disponible = disponible

    def getIntereses(self):
        return self._intereses
    def setIntereses(self, intereses):
        self._intereses = intereses

    def getPrimerMensualidad(self):
        return self._primerMensualidad
    def setPrimerMensualidad(self, primerMensualidad):
        self._primerMensualidad = primerMensualidad

    @classmethod
    def getCuentasCorrienteTotales(cls):
        return cls._cuentasCorrienteTotales
    def setCuentasCorrienteTotales(cls, cuentasCorrienteTotales):
        cls._cuentasCorrienteTotales = cuentasCorrienteTotales